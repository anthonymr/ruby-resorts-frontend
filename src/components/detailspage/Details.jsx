import React, { useEffect } from 'react';
import { useParams, Link, useNavigate } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import {
  Typography, Box, Button, Container,
} from '@mui/material';
import { ArrowCircleRightOutlined, CallEndSharp } from '@mui/icons-material';
import ArrowLeftOutlinedIcon from '@mui/icons-material/ArrowLeftOutlined';
import Rating from './Rating';
import PriceTable from './PriceTable';
import { getRoomDetails } from '../../redux/detailsPage/detailsSlice';
import { useGetRoomDetailQuery } from '../../services/apiService';
import {
  roomName,
  detailsBox,
  reservationprice,
  rightBox,
  ratingBox,
  revervationBtn,
  discoverBtn,
  arrow,
  aprLeft,
  mainBox,
} from './styles';

const Details = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { roomId } = useParams();
  const { data } = useGetRoomDetailQuery(roomId);

  const handleReservationClick = () => {
    navigate(`/newreservepage/${roomId}`);
  };

  useEffect(() => {
    if (data) dispatch(getRoomDetails(data));
  }, [data, dispatch]);

  const { roomDetails } = useSelector((state) => state.details);
  return (
    <Box sx={mainBox}>
      <Container sx={detailsBox}>
        <img
          className="roomImg"
          src={roomDetails.image_url}
          alt="interior of the suite"
        />

        <Box sx={rightBox}>
          <Box
            sx={{
              width: '90%',
            }}
          >
            <Box>
              <Typography variant="h3" sx={roomName}>
                {roomDetails.name}
              </Typography>
              <Typography variant="h6" sx={reservationprice}>
                Full Price $
                {roomDetails.full_price}
              </Typography>
              <PriceTable />
              <Typography sx={aprLeft} varient="h5">
                5.9% APR
                {'  '}
                <span className="details-span">Representative</span>
              </Typography>
            </Box>
          </Box>
          <Box sx={ratingBox}>
            <Link to="/mainpage" className="discoverLink">
              <Typography variant="p" sx={discoverBtn}>
                Discover More Rooms
              </Typography>
              <ArrowCircleRightOutlined sx={arrow} />
            </Link>
            <Rating rating={roomDetails.rating} />
            <Button
              variant="contained"
              color="secondary"
              startIcon={<CallEndSharp />}
              endIcon={<ArrowCircleRightOutlined />}
              sx={revervationBtn}
              onClick={handleReservationClick}
            >
              Book Now
            </Button>
          </Box>
        </Box>
      </Container>
      <button
        type="button"
        className="carousel-button-left back-btn"
        onClick={() => navigate(-1)}
      >
        <ArrowLeftOutlinedIcon />
      </button>
    </Box>
  );
};

export default Details;
