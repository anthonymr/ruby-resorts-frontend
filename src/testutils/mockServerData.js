const userInfoData = {
  id: 1,
  name: 'Test User',
  username: 'testuser',
  email: 'testuser@test.com',
  role: 'admin',
  created_at: '2023-06-17T12:20:06.319Z',
  updated_at: '2023-06-17T12:20:47.606Z',
};

export const roomsData = [
  {
    id: 1,
    name: 'Server Test Suite 1',
    description: 'description of server test suite 1',
    image_url: 'someimageurl1',
  },
  {
    id: 2,
    name: 'Server Test Suite 2',
    description: 'description of server test suite 2',
    image_url: 'someimageurl2',
  },
];

export const hotelsData = [
  {
    id: 1,
    name: 'Test Hotel 1',
    city: 'Test City 1',
    created_at: '2023-06-17T12:15:02.781Z',
    updated_at: '2023-06-17T12:15:02.781Z',
  },
];

export const detailsData = {
  id: 1,
  name: 'Server Test Suite 1',
  description: 'Description of test suite 1',
  full_price: '100.0',
  reservation_price: '10.0',
  reservation_fee: '10.0',
  rating: 5,
  created_at: '2023-06-17T12:15:02.876Z',
  updated_at: '2023-06-17T12:15:02.887Z',
  image_url: 'some image url',
};

export const newRoomData = {
  id: 1,
  name: 'Server New Test Suite',
  description: 'Description of new test suite',
  full_price: '250.0',
  reservation_price: '200.0',
  reservation_fee: '10.0',
  rating: 5,
  created_at: '2023-06-17T12:15:02.876Z',
  updated_at: '2023-06-17T12:15:02.887Z',
  image_url: 'some image url',
};

export const reservationsData = [
  {
    id: 1,
    start_date: '2023-06-18',
    end_date: '2023-06-20',
    user_id: 1,
    hotel_id: 1,
    room_id: 1,
    created_at: '2023-06-18T07:11:15.304Z',
    updated_at: '2023-06-18T07:11:15.304Z',
    amount: '20.0',
    hotel: {
      id: 1,
      name: 'Test Hotel',
      city: 'Test City',
    },
    room: {
      id: 1,
      name: 'Test Suite',
      description:
        'description of test suite 1',
      full_price: '100.0',
      reservation_price: '10.0',
      reservation_fee: '10.0',
      rating: 5,
    },
    user: {
      id: 1,
      name: 'Server Test User 1',
      username: 'testuser',
      email: 'testuser@test.com',
      role: 'admin',
    },
  },
];

export default userInfoData;
