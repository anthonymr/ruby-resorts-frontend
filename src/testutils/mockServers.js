/*eslint-disable*/
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import userInfoData, {
  roomsData,
  hotelsData,
  detailsData,
  reservationsData,
  newRoomData,
} from './mockServerData';

const roomsHandler = [
  rest.get('http://127.0.0.1:3000/api/v1/authentication', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(userInfoData), ctx.delay(10));
  }),
  rest.get('http://127.0.0.1:3000/api/v1/rooms', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(roomsData), ctx.delay(10));
  }),
  rest.get('http://127.0.0.1:3000/api/v1/hotels', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(hotelsData), ctx.delay(1));
  }),
  rest.get('http://127.0.0.1:3000/api/v1/rooms/1', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(detailsData), ctx.delay(1));
  }),
  rest.get('http://127.0.0.1:3000/api/v1/reservations', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(reservationsData), ctx.delay(1));
  }),
  rest.delete('http://127.0.0.1:3000/api/v1/rooms/1', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(detailsData), ctx.delay(1));
  }),
  rest.post('http://127.0.0.1:3000/api/v1/rooms/', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(newRoomData), ctx.delay(1));
  }),
];

export const server1 = setupServer(...roomsHandler);
