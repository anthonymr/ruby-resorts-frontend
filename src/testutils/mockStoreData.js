import { configureStore } from '@reduxjs/toolkit';
import roomsReducer from '../redux/mainPage/roomsSlice';
import detailsReducer from '../redux/detailsPage/detailsSlice';
import citiesReducer from '../redux/newReservePage/citiesSlice';
import userReducer from '../redux/login/userSlice';
import reservationsReducer from '../redux/myReservationsPage/reservationsSlice';
import { appApi } from '../services/apiService';

const initialsStateData = {
  rooms: {
    rooms: [
      {
        id: 1,
        name: 'Test Suite',
        description: 'Description of Test Suite',
        image_url: 'Some URL',
      },
    ],
    status: 'completed',
  },
  cities: {
    hotels: [
      {
        id: 1,
        name: 'Ruby Test Hotel',
        city: 'Test City',
        created_at: '2023-06-16T03:22:52.703Z',
        updated_at: '2023-06-16T03:22:52.703Z',
      },
    ],
  },
  user: {
    accessToken: null,
    userinfo: {
      id: 1,
      name: 'Admin User',
      username: 'admin',
      email: 'test@test.com',
      role: 'admin',
      created_at: '2023-06-16T03:22:52.683Z',
      updated_at: '2023-06-16T03:22:52.683Z',
    },
    error: '',
    authStatus: 'loggedin',
    dataFetched: true,
  },
  appApi: {
    queries: {},
    mutations: {},
    provided: {},
    subscriptions: {},
    config: {
      online: true,
      focused: true,
      middlewareRegistered: false,
      refetchOnFocus: false,
      refetchOnReconnect: false,
      refetchOnMountOrArgChange: false,
      keepUnusedDataFor: 60,
      reducerPath: 'appApi',
    },
  },
  reservations: {
    reservations: [
      {
        id: 1,
        start_date: '2023-06-16',
        end_date: '2023-06-17',
        user_id: 1,
        hotel_id: 1,
        room_id: 2,
        created_at: '2023-06-16T05:26:03.849Z',
        updated_at: '2023-06-16T05:26:03.849Z',
        amount: '220.0',
        hotel: {
          id: 1,
          name: 'Ruby Amsterdam',
          city: 'Amsterdam',
        },
        room: {
          id: 2,
          name: 'Emerald',
          description:
            'Emerald Suites are second only to our Ruby Suites when it comes to luxury. The suites come with 5-star amenities',
          full_price: '250.0',
          reservation_price: '220.0',
          reservation_fee: '5.0',
          rating: 5,
        },
        user: {
          id: 1,
          name: 'Admin User',
          username: 'admin',
          email: 'test@test.com',
          role: 'admin',
        },
      },
    ],
    status: 'completed',
  },
};

export const setMockStore = (preloadedState) => configureStore({
  reducer: {
    rooms: roomsReducer,
    cities: citiesReducer,
    user: userReducer,
    [appApi.reducerPath]: appApi.reducer,
    details: detailsReducer,
    reservations: reservationsReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware({
    serializableCheck: false,
  }).concat(appApi.middleware),
  preloadedState,
});

export default initialsStateData;
