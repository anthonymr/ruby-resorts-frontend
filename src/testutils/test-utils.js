import React from 'react';
import PropTypes from 'prop-types';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../redux/store';

const render = (ui, options) => {
  const Wrapper = ({ children }) => <Provider store={store}>{children}</Provider>;
  Wrapper.propTypes = {
    children: PropTypes.node.isRequired,
  };

  return rtlRender(ui, { wrapper: Wrapper, ...options });
};

export * from '@testing-library/react';
export { render };
