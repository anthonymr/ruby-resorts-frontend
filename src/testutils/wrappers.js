import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { PropTypes } from 'prop-types';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import initialStateData, { setMockStore } from './mockStoreData';

export function renderWithProviders(
  ui,
  {
    preloadedState = initialStateData,
    store = setMockStore(preloadedState),
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    Wrapper.propTypes = {
      children: PropTypes.node.isRequired,
    };
    return <Provider store={store}>{children}</Provider>;
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export const renderWithRouter = (
  ui,
  {
    preloadedState = initialStateData,
    store = setMockStore(preloadedState),
    ...renderOptions
  } = {},
) => {
  function Wrapper({ children }) {
    Wrapper.propTypes = {
      children: PropTypes.node.isRequired,
    };
    return (
      <MemoryRouter initialEntries={['/newreservepage/0']}>
        <Provider store={store}>
          <Routes>
            <Route path="/newreservepage/:roomId" element={children} />
          </Routes>
        </Provider>
      </MemoryRouter>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};
