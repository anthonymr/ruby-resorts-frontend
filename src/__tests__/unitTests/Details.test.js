import { BrowserRouter as Router } from 'react-router-dom';
import { render } from '../../testutils/test-utils';
import Details from '../../components/detailspage/Details';

it('renders correctly', () => {
  const { asFragment } = render(
    <Router>
      <Details />
    </Router>,
  );
  expect(asFragment()).toMatchSnapshot();
});
