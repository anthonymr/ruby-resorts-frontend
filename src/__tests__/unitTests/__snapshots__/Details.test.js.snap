// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1ezxzkz"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-16chdws-MuiContainer-root"
    >
      <img
        alt="interior of the suite"
        class="roomImg"
      />
      <div
        class="MuiBox-root css-1rbp01k"
      >
        <div
          class="MuiBox-root css-7822g1"
        >
          <div
            class="MuiBox-root css-0"
          >
            <h3
              class="MuiTypography-root MuiTypography-h3 css-125ornw-MuiTypography-root"
            />
            <h6
              class="MuiTypography-root MuiTypography-h6 css-dppiby-MuiTypography-root"
            >
              Full Price $
            </h6>
            <div
              class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
            >
              <table
                class="MuiTable-root css-gh8rj1-MuiTable-root"
              >
                <tbody
                  class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
                >
                  <tr
                    class="MuiTableRow-root css-kxwusg-MuiTableRow-root"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1y7quc6-MuiTableCell-root"
                      scope="row"
                    >
                      Reservation Price
                    </th>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-f9hqh2-MuiTableCell-root"
                    />
                  </tr>
                  <tr
                    class="MuiTableRow-root css-kxwusg-MuiTableRow-root"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1y7quc6-MuiTableCell-root"
                      scope="row"
                    >
                      Transaction fee
                    </th>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-f9hqh2-MuiTableCell-root"
                    />
                  </tr>
                  <tr
                    class="MuiTableRow-root css-kxwusg-MuiTableRow-root"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1y7quc6-MuiTableCell-root"
                      scope="row"
                    >
                      Total amount
                    </th>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-f9hqh2-MuiTableCell-root"
                    >
                      0
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root css-kxwusg-MuiTableRow-root"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1y7quc6-MuiTableCell-root"
                      scope="row"
                    >
                      Cancelation Fee
                    </th>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-f9hqh2-MuiTableCell-root"
                    >
                      30%
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1495mji-MuiTypography-root"
              varient="h5"
            >
              5.9% APR  
              <span
                class="details-span"
              >
                Representative
              </span>
            </p>
          </div>
        </div>
        <div
          class="MuiBox-root css-11yvj5x"
        >
          <a
            class="discoverLink"
            href="/mainpage"
          >
            <span
              class="MuiTypography-root MuiTypography-p css-13r7zsy-MuiTypography-root"
            >
              Discover More Rooms
            </span>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-7ksnoo-MuiSvgIcon-root"
              data-testid="ArrowCircleRightOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M22 12c0-5.52-4.48-10-10-10S2 6.48 2 12s4.48 10 10 10 10-4.48 10-10zM4 12c0-4.42 3.58-8 8-8s8 3.58 8 8-3.58 8-8 8-8-3.58-8-8zm12 0-4 4-1.41-1.41L12.17 13H8v-2h4.17l-1.59-1.59L12 8l4 4z"
              />
            </svg>
          </a>
          <div
            class="MuiBox-root css-167yxyd"
          >
            <div
              data-test-id="CircularProgressbarWithChildren"
            >
              <div
                style="position: relative; width: 100%; height: 100%;"
              >
                <svg
                  class="CircularProgressbar "
                  data-test-id="CircularProgressbar"
                  viewBox="0 0 100 100"
                >
                  <path
                    class="CircularProgressbar-trail"
                    d="
      M 50,50
      m 0,-45
      a 45,45 0 1 1 0,90
      a 45,45 0 1 1 0,-90
    "
                    fill-opacity="0"
                    stroke-width="10"
                    style="stroke-linecap: butt; stroke-dasharray: 282.7433388230814px 282.7433388230814px; stroke-dashoffset: 0px;"
                  />
                  <path
                    class="CircularProgressbar-path"
                    d="
      M 50,50
      m 0,-45
      a 45,45 0 1 1 0,90
      a 45,45 0 1 1 0,-90
    "
                    fill-opacity="0"
                    stroke-width="10"
                    style="stroke: #96bf01; stroke-linecap: butt; stroke-dasharray: 282.7433388230814px 282.7433388230814px; stroke-dashoffset: 0px;"
                  />
                  <text
                    class="CircularProgressbar-text"
                    style="fill: #96bf01;"
                    x="50"
                    y="50"
                  >
                    5
                  </text>
                </svg>
                <div
                  data-test-id="CircularProgressbarWithChildren__children"
                  style="position: absolute; width: 100%; height: 100%; margin-top: -100%; display: flex; flex-direction: column; justify-content: center; align-items: center;"
                >
                  <div
                    style="position: absolute; height: 100%; transform: rotate(0turn);"
                  >
                    <div
                      style="background: rgb(255, 255, 255); width: 3px; height: 10%;"
                    />
                  </div>
                  <div
                    style="position: absolute; height: 100%; transform: rotate(0.2turn);"
                  >
                    <div
                      style="background: rgb(255, 255, 255); width: 3px; height: 10%;"
                    />
                  </div>
                  <div
                    style="position: absolute; height: 100%; transform: rotate(0.4turn);"
                  >
                    <div
                      style="background: rgb(255, 255, 255); width: 3px; height: 10%;"
                    />
                  </div>
                  <div
                    style="position: absolute; height: 100%; transform: rotate(0.6000000000000001turn);"
                  >
                    <div
                      style="background: rgb(255, 255, 255); width: 3px; height: 10%;"
                    />
                  </div>
                  <div
                    style="position: absolute; height: 100%; transform: rotate(0.8turn);"
                  >
                    <div
                      style="background: rgb(255, 255, 255); width: 3px; height: 10%;"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1qssj0y-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CallEndSharpIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m3.68 16.07 3.92-3.11V9.59c2.85-.93 5.94-.93 8.8 0v3.38l3.91 3.1L24 12.39c-6.41-7.19-17.59-7.19-24 0l3.68 3.68z"
                />
              </svg>
            </span>
            Book Now
            <span
              class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowCircleRightOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22 12c0-5.52-4.48-10-10-10S2 6.48 2 12s4.48 10 10 10 10-4.48 10-10zM4 12c0-4.42 3.58-8 8-8s8 3.58 8 8-3.58 8-8 8-8-3.58-8-8zm12 0-4 4-1.41-1.41L12.17 13H8v-2h4.17l-1.59-1.59L12 8l4 4z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <button
      class="carousel-button-left back-btn"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ArrowLeftOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m14 7-5 5 5 5V7z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
